### PR TITLE
Fix deletion of resumed checkpoints

### LINKED ISSUE #16426: ModelCheckpoint(every_n_train_steps=...) and loading from checkpoint deletes loaded ckpt file
### Bug description

If a model is restarted from a checkpoint file e.g `trainer.fit(...., ckpt_path="prev_version/abc.cpkt")` and the Trainer callback `ModelCheckpoint(every_n_train_steps=n)` is called, at step n the old checkpoint is deleted and not the one in current log_dir. This is usually a disaster if a checkpoint file is deleted. 

### How to reproduce the bug

```python
ckpt_callback = ModelCheckpoint(every_n_train_steps=n)

trainer = pl.Trainer(
                     ...
                     callbacks=[ckpt_callback],
                     ...
)

trainer.fit(model, datamodule=dm, ckpt_path=somewhere/checkpoint.ckpt)

# somewhere/checkpoint.ckpt is deleted after n steps and not a checkpoint file in the current log_dir
```
```


### Error messages and logs

```
# I removed write permission on the old checkpoint directory and therefore an error occurs, of course:
File "/home/mheusel/.conda/envs/my/lib/python3.10/site-packages/pytorch_lightning/strategies/launchers/subprocess_script.py", line 90, in launch
   return function(*args, **kwargs)
 File "/home/mheusel/.conda/envs/my/lib/python3.10/site-packages/pytorch_lightning/trainer/trainer.py", line 624, in _fit_impl
   self._run(model, ckpt_path=self.ckpt_path)
 File "/home/mheusel/.conda/envs/my/lib/python3.10/site-packages/pytorch_lightning/trainer/trainer.py", line 1061, in _run
   results = self._run_stage()
 File "/home/mheusel/.conda/envs/my/lib/python3.10/site-packages/pytorch_lightning/trainer/trainer.py", line 1140, in _run_stage
   self._run_train()
 File "/home/mheusel/.conda/envs/my/lib/python3.10/site-packages/pytorch_lightning/trainer/trainer.py", line 1163, in _run_train
   self.fit_loop.run()
 File "/home/mheusel/.conda/envs/my/lib/python3.10/site-packages/pytorch_lightning/loops/loop.py", line 199, in run
   self.advance(*args, **kwargs)
 File "/home/mheusel/.conda/envs/my/lib/python3.10/site-packages/pytorch_lightning/loops/fit_loop.py", line 267, in advance
   self._outputs = self.epoch_loop.run(self._data_fetcher)
 File "/home/mheusel/.conda/envs/my/lib/python3.10/site-packages/pytorch_lightning/loops/loop.py", line 199, in run
   self.advance(*args, **kwargs)
 File "/home/mheusel/.conda/envs/my/lib/python3.10/site-packages/pytorch_lightning/loops/epoch/training_epoch_loop.py", line 230, in advance
   self.trainer._call_callback_hooks("on_train_batch_end", batch_end_outputs, batch, batch_idx)
 File "/home/mheusel/.conda/envs/my/lib/python3.10/site-packages/pytorch_lightning/trainer/trainer.py", line 1343, in _call_callback_hooks
   fn(self, self.lightning_module, *args, **kwargs)
 File "/home/mheusel/.conda/envs/my/lib/python3.10/site-packages/pytorch_lightning/callbacks/model_checkpoint.py", line 304, in on_train_batch_end
   self._save_topk_checkpoint(trainer, monitor_candidates)
 File "/home/mheusel/.conda/envs/my/lib/python3.10/site-packages/pytorch_lightning/callbacks/model_checkpoint.py", line 371, in
_save_topk_checkpoint
   self._save_none_monitor_checkpoint(trainer, monitor_candidates)
 File "/home/mheusel/.conda/envs/my/lib/python3.10/site-packages/pytorch_lightning/callbacks/model_checkpoint.py", line 659, in
_save_none_monitor_checkpoint
   trainer.strategy.remove_checkpoint(previous)
 File "/home/mheusel/.conda/envs/my/lib/python3.10/site-packages/pytorch_lightning/strategies/strategy.py", line 475, in remove_checkpoint
   self.checkpoint_io.remove_checkpoint(filepath)
 File "/home/mheusel/.conda/envs/my/lib/python3.10/site-packages/lightning_lite/plugins/io/torch_io.py", line 96, in remove_checkpoint
   fs.rm(path, recursive=True)
 File "/home/mheusel/.conda/envs/my/lib/python3.10/site-packages/fsspec/implementations/local.py", line 169, in rm
   os.remove(p)
PermissionError: [Errno 13] Permission denied:
'..../lightning_logs/version_137/checkpoints/epoch=3-step=118000.ckpt'

Epoch 3/3  ━╺━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1000/37323 0:58:23 • 1 day, 11:04:50 0.29it/s v_num: 144

```


### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 1.10):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>


### More info

_No response_

### LINKED ISSUE #17189: Disable dangerous checkpoint deletion behaviour
### Description & Motivation

When PL resumes training from a checkpoint, it deletes the old checkpoint from the version directory. Over the last several weeks, this behaviour has led to loss of several days of training:
- resume old run, NaN loss occurs in new run -> new run checkpoint unusable, but old checkpoint has been removed -> have to start training from scratch.
- new run contains problematic samples -> new run checkpoint unusable, but old checkpoint removed -> have to start training from scratch.
- out of disk space -> new run checkpoint fails to write, but old checkpoint already deleted -> have to start training from scratch.
- pretrained model -> want to finetune several variations, but PL keeps deleting the pretrained checkpoint -> have to work around PL to achieve this.

### Pitch

Is there a way to disable this behaviour or at least make it opt-in? (I did not find it in the documentation.)

The value of a checkpoint that took weeks to train is MUCH higher than saving a few hundred MB of disk space by deleting resumed checkpoints.

### Alternatives

I have resorted to making copies of checkpoints manually via the filesystem, but this is error prone and time-consuming.

### Additional context

_No response_

cc @borda @tchaton @justusschock @awaelchli @carmocca

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #17189
Fixes #16426

This change makes it so that ModelCheckpoint callback no longer deletes a checkpoint if it originates from a directory that is not a subpath of the ModelCheckpoint's dirpath. This is to prevent accidental deletion of checkpoints when resuming. Furthermore, we now also guarantee that the checkpoint passed to the `trainer.fit(..., ckpt_path=)` never gets deleted, even if it would be subject to the save-top-k mechanism.

----
:books: Documentation preview :books:: https://pytorch-lightning--18750.org.readthedocs.build/en/18750/

cc @borda @carmocca @awaelchli

### CODE REVIEW

File: tests/tests_pytorch/checkpointing/test_model_checkpoint.py [hunk 1]
```diff
@@ -1420,3 +1420,31 @@ def test_train_epoch_end_ckpt_with_no_validation():
     assert not trainer.checkpoint_callback._should_save_on_train_epoch_end(trainer)
     trainer.val_check_interval = 0.8
     assert not trainer.checkpoint_callback._should_save_on_train_epoch_end(trainer)
+
+
+def test_resume_and_old_checkpoint_files_remain(tmp_path):
+    """Test that checkpoints saved in the resume-folder won't be delteded under the save-top-k mechanism."""
```
Comment (awaelchli) [inline tests/tests_pytorch/checkpointing/test_model_checkpoint.py:1426]:
```suggestion
    """Test that checkpoints saved in the resume-folder won't be deleted under the save-top-k mechanism."""
```

File: src/lightning/pytorch/callbacks/model_checkpoint.py [hunk 2]
```diff
@@ -742,6 +744,27 @@ def file_exists(self, filepath: _PATH, trainer: "pl.Trainer") -> bool:
         exists = self._fs.exists(filepath)
         return trainer.strategy.broadcast(exists)
 
+    def _should_remove_checkpoint(self, trainer: "pl.Trainer", previous: str, current: str) -> bool:
+        """Checks if the previous checkpoint should be deleted.
+
+        A checkpoint won't be deleted if it is a local file and any of the following cases apply:
+        - The previous checkpoint is the same as the current checkpoint
```
Comment (carmocca) [inline src/lightning/pytorch/callbacks/model_checkpoint.py:751]: This is also a condition when it is not a local file. The docstring makes it sound like this only applies to local files

File: src/lightning/pytorch/callbacks/model_checkpoint.py [hunk 3]
```diff
@@ -742,6 +744,27 @@ def file_exists(self, filepath: _PATH, trainer: "pl.Trainer") -> bool:
...
+
+        """
+        if previous == current:
+            return False
+        if self._fs.protocol != "file":
+            return True
+        previous = Path(previous).absolute()
+        resume_path = Path(trainer.ckpt_path).absolute() if trainer.ckpt_path is not None else None
+        if resume_path is not None and previous == resume_path:
+            return False
```
Comment (carmocca) [inline src/lightning/pytorch/callbacks/model_checkpoint.py:763]: This should be equivalent

```suggestion
        if trainer.ckpt_path is None:
            return False
        previous = Path(previous).absolute()
        if previous == Path(trainer.ckpt_path).absolute():
            return False
```

**Review by carmocca — APPROVED** [review]

File: src/lightning/pytorch/callbacks/model_checkpoint.py (hunk 3 above)
Comment (awaelchli) [reply src/lightning/pytorch/callbacks/model_checkpoint.py]: Oh I was too quick, actually no it is not equivalent. With this code, we would return False too early under the condition `if trainer.ckpt_path is None`. We would still have to check the conditions below. The previous version was correct.

File: src/lightning/pytorch/callbacks/model_checkpoint.py (hunk 2 above)
Comment (awaelchli) [reply src/lightning/pytorch/callbacks/model_checkpoint.py:751]: Thanks for spotting it. I refined the description

File: src/lightning/pytorch/callbacks/model_checkpoint.py [hunk 4]
```diff
@@ -742,6 +744,27 @@ def file_exists(self, filepath: _PATH, trainer: "pl.Trainer") -> bool:
...
+        assert self.dirpath is not None
```
Comment (Borda) [inline src/lightning/pytorch/callbacks/model_checkpoint.py:764]:
```suggestion
        if self.dirpath is None:
            raise IsADirectoryError("Attribute `dirpath` is not set.")
```

**Review by Borda — APPROVED** [review]

**Pull request merged**